### PR TITLE
DfciPkg/DfciMenu: Fix memory leaks

### PR DESCRIPTION
**Pull request by wenbhou**
## Description

Fix https://github.com/microsoft/mu_feature_dfci/issues/65

- [ ] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Run standard testing:

1. DFCI_InitialState
2. DFCI_InTuneEnroll
3. DFCI_InTuneRollCerts
4. DFCI_InTunePermissions
5. DFCI_InTuneSettings
6. DFCI_InTuneBadUpdate
7. DFCI_InTuneUnenroll

## Integration Instructions

N/A

### CODE REVIEW

File: DfciPkg/Application/DfciMenu/DfciRequest.c [hunk 1]
```diff
@@ -1193,7 +1193,7 @@ ProcessHttpRequest (
              &RequestMessage.HeaderCount
              );
   if (EFI_ERROR (Status)) {
-    return Status;
+    goto S_EXIT1;
```
[the remark this answers is not in the document]
Comment (wenbhou) [reply DfciPkg/Application/DfciMenu/DfciRequest.c:1196]: Here to goto S_EXIT1 is to free the RequestData.Url which is already allocated from DfciConvertToCHAR16()

[the remark this answers is not in the document]
Comment (makubacki) [reply DfciPkg/Application/DfciMenu/DfciRequest.c:1196]: I'm talking about this call in `S_EXIT1`:

```c
HttpFreeHeaderFields (RequestMessage.Headers, RequestMessage.HeaderCount);
```

Who else would be freeing the buffers allocated in `DfciBuildRequestHeaders()`?

File: DfciPkg/Application/DfciMenu/DfciUpdate.c
[the remark this answers is not in the document]
Comment (wenbhou) [reply DfciPkg/Application/DfciMenu/DfciUpdate.c:1]: Reverted.

File: DfciPkg/Application/DfciMenu/DfciRequest.c
Comment (makubacki) [inline DfciPkg/Application/DfciMenu/DfciRequest.c:1]: Another problem is that the blanket checking of allocations being null will further miss freeing individual allocations that succeeded in the way the function is currently written.

```c
  if ((RequestHeaders[0].FieldName == NULL) ||
      (RequestHeaders[1].FieldName == NULL) ||
      (RequestHeaders[1].FieldValue == NULL) ||
      (RequestHeaders[2].FieldName == NULL) ||
      (RequestHeaders[2].FieldValue == NULL) ||
      (RequestHeaders[3].FieldName == NULL) ||
      (RequestHeaders[3].FieldValue == NULL))
  {
    return EFI_OUT_OF_RESOURCES;
  }
```

Comment (makubacki) [reply DfciPkg/Application/DfciMenu/DfciRequest.c:1]: Thanks for addressing this in the latest commit

File: DfciPkg/Application/DfciMenu/DfciRequest.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (makubacki) [reply DfciPkg/Application/DfciMenu/DfciRequest.c:1208]: Thanks for addressing this in the latest commit.

**Review by makubacki — APPROVED** [review]

**Review by apop5 — APPROVED** [review]

**Pull request merged**